pkg/auth/azure_test.go

```diff
@@ -0,0 +1,231 @@
+/*This file is part of kuberpult.
+
+Kuberpult is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+Kuberpult is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with kuberpult.  If not, see <http://www.gnu.org/licenses/>.
+
+Copyright 2021 freiheit.com*/
+package auth
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/MicahParks/keyfunc"
+	jwt "github.com/golang-jwt/jwt/v4"
+	"github.com/google/go-cmp/cmp"
+)
+
+func TestValidateTokenStatic(t *testing.T) {
+	tcs := []struct {
+		Name          string
+		Token         string
+		ExpectedError string
+		noInit        bool
+	}{
+		{
+			Name:          "Not a token",
+			Token:         "asdf",
+			ExpectedError: "Failed to parse the JWT.\nError: token contains an invalid number of segments",
+		},
+		{
+			Name:          "Not initialized",
+			Token:         "asdf",
+			noInit:        true,
+			ExpectedError: "JWKS not initialized.",
+		},
+		{
+			Name:          "Not a token 2",
+			Token:         "asdf.asdf.asdf",
+			ExpectedError: "Failed to parse the JWT.\nError: invalid character 'j' looking for beginning of value",
+		},
+		{
+			Name:          "Kid not present",
+			Token:         "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.WDlNbJFe8ZX6C1mS27xwxg-9tk8vtkk6sDgucRj8xW0",
+			ExpectedError: "Failed to parse the JWT.\nError: the JWT has an invalid kid: could not find kid in JWT header",
+		},
+		{
+			Name:          "Kid not part of jwks",
+			Token:         "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6ImFzZGYifQ.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.aNyAK8qpCScGchUmv1q1pBXOddWKN8_7agLUo7pXDog",
+			ExpectedError: "Failed to parse the JWT.\nError: the given key ID was not found in the JWKS",
+		},
+	}
+
+	for _, tc := range tcs {
+		tc := tc
+		t.Run(tc.Name, func(t *testing.T) {
+			t.Parallel()
+			ctx := context.Background()
+			var jwks *keyfunc.JWKS
+			var err error
+			if !(tc.noInit) {
+				jwks, err = JWKSInitAzure(ctx)
```

Comment:
This means we are doing a several requests to `https://login.microsoftonline.com/common/discovery/v2.0/keys` right?
We might run into some sort of rate limiting. That could be dangerous, as our build would randomly fail.

Reply:
I would suggest that we store the keys, instead of requesting them again on each test.

Reply:
updating in follow up pr